### DISCUSSION TITLE
[Issue] Translation updates take too long

### DISCUSSION THREAD

**Discussion by chennes** [FreeCAD/FreeCAD#10198] [Discussions that will turn in to issues but need refining]
### Issue description

In PR #9275 @ronnystandtke points out that our current translation workflow introduces a very long lag between when a string is made translatable in the source code, and when that string is actually available in translated form.

I think there are two things at work here:
1. The roundtrip from source code to CrowdIn, and vice versa, is entirely manual right now, and is not done on a regular schedule. This means that developers and translators never know when to expect their work to appear.
2. There is not good documentation on how developers can immediately test that their newly-translatable strings can in fact be translated. I believe this is done by putting translation files in $USER_APP_DATA_DIR/languages.

**Comment by adrianinsaval** [comment]
How is the upload to crowdin made? Is there mechanism to have a token to authenticate and automate the process? If so it could be done on a schedule via github actions.

It might also be possible to make an action to generate a PR for updating the translations in the repo, but that might clutter the repo with useless branches.

**Comment by d1nuc0m** [comment]
> Is there mechanism to have a token to authenticate and automate the process?

These may be interesting:

- [GitHub Crowdin Action [GitHub]](https://github.com/crowdin/github-action)
- [GitHub Actions [Crowdin]](https://store.crowdin.com/github-actions-webhook)

**Comment by chennes** [comment]
I'm ultimately going to start automating this with CrowdIn's GitHub actions, but before I do I want to integrate the generation of the QM and QRC files into the build process. I think we can get that working in a couple of weeks here.

**Comment by yorikvanhavre** [comment]
We definitely need to remove qm files from the source before ramping up translations frequency, otherwise it will just inflate the repo size dramatically.

Also, there is the issue that when we use another version of lupdate, it sometimes rewrite all the ts files (changes the order somehow). This is less critical of course, but might be a good idea to define a version to be used and allow no other...

> How is the upload to crowdin made? Is there mechanism to have a token to authenticate and automate the process? If so it could be done on a schedule via github actions.

Right now, it's all made manually with src/Tools/updateCrowdin.py. It has different commands to gather strings into "master" ts files, upload these ts files to crowdin, then build, download and apply the translated ts files.

Anyone with manager rights on the crowdin repo can run the script (you need to store an API key on your machine). I suppose it could be ran from github, but one would have to expose their personal crowdin API key which seems not a good idea.

Maybe what could be done is to have a github action to gather and build the master ts files only. Than maybe crowdin has some system to automatically come fetch those files?

**Comment by chennes** [comment]
CrowdIn has pretty good GitHub integration, we can actually go both directions if we want (they can automatically generate PRs with the new TS files). I'm looking into how it all works.
